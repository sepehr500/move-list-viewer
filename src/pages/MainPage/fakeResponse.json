{
  "page": 1,
  "total_results": 15631,
  "total_pages": 782,
  "results": [
    {
      "vote_count": 8339,
      "id": 246655,
      "video": false,
      "vote_average": 6.5,
      "title": "X-Men: Apocalypse",
      "popularity": 51.257,
      "poster_path": "/zSouWWrySXshPCT4t3UKCQGayyo.jpg",
      "original_language": "en",
      "original_title": "X-Men: Apocalypse",
      "genre_ids": [
        28,
        12,
        878,
        14
      ],
      "backdrop_path": "/oQWWth5AOtbWG9o8SCAviGcADed.jpg",
      "adult": false,
      "overview": "After the re-emergence of the world's first mutant, world-destroyer Apocalypse, the X-Men must unite to defeat his extinction level plan.",
      "release_date": "2016-05-18"
    },
    {
      "vote_count": 13991,
      "id": 271110,
      "video": false,
      "vote_average": 7.4,
      "title": "Captain America: Civil War",
      "popularity": 39.833,
      "poster_path": "/kSBXou5Ac7vEqKd97wotJumyJvU.jpg",
      "original_language": "en",
      "original_title": "Captain America: Civil War",
      "genre_ids": [
        12,
        28,
        878
      ],
      "backdrop_path": "/yi3hZ4mQq8JGyJSi3MBW5nMj1f.jpg",
      "adult": false,
      "overview": "Following the events of Age of Ultron, the collective governments of the world pass an act designed to regulate all superhuman activity. This polarizes opinion amongst the Avengers, causing two factions to side with Iron Man or Captain America, which causes an epic battle between former allies.",
      "release_date": "2016-04-27"
    },
    {
      "vote_count": 20164,
      "id": 293660,
      "video": false,
      "vote_average": 7.6,
      "title": "Deadpool",
      "popularity": 36.218,
      "poster_path": "/inVq3FRqcYIRl2la8iZikYYxFNR.jpg",
      "original_language": "en",
      "original_title": "Deadpool",
      "genre_ids": [
        28,
        12,
        35
      ],
      "backdrop_path": "/n1y094tVDFATSzkTnFxoGZ1qNsG.jpg",
      "adult": false,
      "overview": "Deadpool tells the origin story of former Special Forces operative turned mercenary Wade Wilson, who after being subjected to a rogue experiment that leaves him with accelerated healing powers, adopts the alter ego Deadpool. Armed with his new abilities and a dark, twisted sense of humor, Deadpool hunts down the man who nearly destroyed his life.",
      "release_date": "2016-02-09"
    },
    {
      "vote_count": 12142,
      "id": 259316,
      "video": false,
      "vote_average": 7.4,
      "title": "Fantastic Beasts and Where to Find Them",
      "popularity": 34.897,
      "poster_path": "/1M91Bt3oGspda75H9eLqYZkJzgO.jpg",
      "original_language": "en",
      "original_title": "Fantastic Beasts and Where to Find Them",
      "genre_ids": [
        12,
        10751,
        14
      ],
      "backdrop_path": "/7GrpqAs0oDcFcwFwyygnUI7BrZA.jpg",
      "adult": false,
      "overview": "In 1926, Newt Scamander arrives at the Magical Congress of the United States of America with a magically expanded briefcase, which houses a number of dangerous creatures and their habitats. When the creatures escape from the briefcase, it sends the American wizarding authorities after Newt, and threatens to strain even further the state of magical and non-magical relations.",
      "release_date": "2016-11-16"
    },
    {
      "vote_count": 11988,
      "id": 209112,
      "video": false,
      "vote_average": 5.8,
      "title": "Batman v Superman: Dawn of Justice",
      "popularity": 33.203,
      "poster_path": "/cGOPbv9wA5gEejkUN892JrveARt.jpg",
      "original_language": "en",
      "original_title": "Batman v Superman: Dawn of Justice",
      "genre_ids": [
        28,
        12,
        14
      ],
      "backdrop_path": "/vsjBeMPZtyB7yNsYY56XYxifaQZ.jpg",
      "adult": false,
      "overview": "Fearing the actions of a god-like Super Hero left unchecked, Gotham City’s own formidable, forceful vigilante takes on Metropolis’s most revered, modern-day savior, while the world wrestles with what sort of hero it really needs. And with Batman and Superman at war with one another, a new threat quickly arises, putting mankind in greater danger than it’s ever known before.",
      "release_date": "2016-03-23"
    },
    {
      "vote_count": 12575,
      "id": 284052,
      "video": false,
      "vote_average": 7.3,
      "title": "Doctor Strange",
      "popularity": 32.022,
      "poster_path": "/4PiiNGXj1KENTmCBHeN6Mskj2Fq.jpg",
      "original_language": "en",
      "original_title": "Doctor Strange",
      "genre_ids": [
        28,
        12,
        14,
        878
      ],
      "backdrop_path": "/tFI8VLMgSTTU38i8TIsklfqS9Nl.jpg",
      "adult": false,
      "overview": "After his career is destroyed, a brilliant but arrogant surgeon gets a new lease on life when a sorcerer takes him under her wing and trains him to defend the world against evil.",
      "release_date": "2016-10-25"
    },
    {
      "vote_count": 9967,
      "id": 269149,
      "video": false,
      "vote_average": 7.7,
      "title": "Zootopia",
      "popularity": 30.416,
      "poster_path": "/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg",
      "original_language": "en",
      "original_title": "Zootopia",
      "genre_ids": [
        16,
        12,
        10751,
        35
      ],
      "backdrop_path": "/mhdeE1yShHTaDbJVdWyTlzFvNkr.jpg",
      "adult": false,
      "overview": "Determined to prove herself, Officer Judy Hopps, the first bunny on Zootopia's police force, jumps at the chance to crack her first case - even if it means partnering with scam-artist fox Nick Wilde to solve the mystery.",
      "release_date": "2016-02-11"
    },
    {
      "vote_count": 6739,
      "id": 291805,
      "video": false,
      "vote_average": 6.8,
      "title": "Now You See Me 2",
      "popularity": 28.238,
      "poster_path": "/hU0E130tsGdsYa4K9lc3Xrn5Wyt.jpg",
      "original_language": "en",
      "original_title": "Now You See Me 2",
      "genre_ids": [
        28,
        35,
        53
      ],
      "backdrop_path": "/zrAO2OOa6s6dQMQ7zsUbDyIBrAP.jpg",
      "adult": false,
      "overview": "One year after outwitting the FBI and winning the public’s adulation with their mind-bending spectacles, the Four Horsemen resurface only to find themselves face to face with a new enemy who enlists them to pull off their most dangerous heist yet.",
      "release_date": "2016-06-02"
    },
    {
      "vote_count": 4151,
      "id": 372058,
      "video": false,
      "vote_average": 8.6,
      "title": "Your Name.",
      "popularity": 24.527,
      "poster_path": "/xq1Ugd62d23K2knRUx6xxuALTZB.jpg",
      "original_language": "ja",
      "original_title": "君の名は。",
      "genre_ids": [
        10749,
        16,
        18
      ],
      "backdrop_path": "/nvsdKYPKwf51EAgC0xLJMB9kUZM.jpg",
      "adult": false,
      "overview": "High schoolers Mitsuha and Taki are complete strangers living separate lives. But one night, they suddenly switch places. Mitsuha wakes up in Taki’s body, and he in hers. This bizarre occurrence continues to happen randomly, and the two must adjust their lives around each other.",
      "release_date": "2016-08-26"
    },
    {
      "vote_count": 6601,
      "id": 324786,
      "video": false,
      "vote_average": 8.1,
      "title": "Hacksaw Ridge",
      "popularity": 22.406,
      "poster_path": "/bndiUFfJxNd2fYx8XO610L9a07m.jpg",
      "original_language": "en",
      "original_title": "Hacksaw Ridge",
      "genre_ids": [
        18,
        36,
        10752
      ],
      "backdrop_path": "/zBK4QZONMQXhcgaJv1YYTdCW7q9.jpg",
      "adult": false,
      "overview": "WWII American Army Medic Desmond T. Doss, who served during the Battle of Okinawa, refuses to kill people and becomes the first Conscientious Objector in American history to receive the Congressional Medal of Honor.",
      "release_date": "2016-10-07"
    },
    {
      "vote_count": 3828,
      "id": 334543,
      "video": false,
      "vote_average": 8.1,
      "title": "Lion",
      "popularity": 22.368,
      "poster_path": "/iBGRbLvg6kVc7wbS8wDdVHq6otm.jpg",
      "original_language": "en",
      "original_title": "Lion",
      "genre_ids": [
        18
      ],
      "backdrop_path": "/fgffluhvdOv79SkYaKrfRA0xkf4.jpg",
      "adult": false,
      "overview": "A five-year-old Indian boy gets lost on the streets of Calcutta, thousands of kilometers from home. He survives many challenges before being adopted by a couple in Australia; 25 years later, he sets out to find his lost family.",
      "release_date": "2016-11-24"
    },
    {
      "vote_count": 497,
      "id": 348893,
      "video": false,
      "vote_average": 6.2,
      "title": "Boyka: Undisputed IV",
      "popularity": 22.198,
      "poster_path": "/mWiqc87iTs1qujxm2Q0NRzGvWSN.jpg",
      "original_language": "en",
      "original_title": "Boyka: Undisputed IV",
      "genre_ids": [
        53,
        28,
        80
      ],
      "backdrop_path": "/5dLLiR8IUT7T0r5HVcvRlItNWTn.jpg",
      "adult": false,
      "overview": "In the fourth installment of the fighting franchise, Boyka is shooting for the big leagues when an accidental death in the ring makes him question everything he stands for. When he finds out the wife of the man he accidentally killed is in trouble, Boyka offers to fight in a series of impossible battles to free her from a life of servitude",
      "release_date": "2016-08-01"
    },
    {
      "vote_count": 2773,
      "id": 311324,
      "video": false,
      "vote_average": 5.8,
      "title": "The Great Wall",
      "popularity": 21.682,
      "poster_path": "/p70dq1YxabemdZDm5K6Q8G10wSn.jpg",
      "original_language": "en",
      "original_title": "The Great Wall",
      "genre_ids": [
        28,
        12,
        14
      ],
      "backdrop_path": "/rym1ecQJVj1NhZnigRePtiusorF.jpg",
      "adult": false,
      "overview": "European mercenaries searching for black powder become embroiled in the defense of the Great Wall of China against a horde of monstrous creatures.",
      "release_date": "2016-12-16"
    },
    {
      "vote_count": 5254,
      "id": 278927,
      "video": false,
      "vote_average": 6.8,
      "title": "The Jungle Book",
      "popularity": 21.458,
      "poster_path": "/vOipe2myi26UDwP978hsYOrnUWC.jpg",
      "original_language": "en",
      "original_title": "The Jungle Book",
      "genre_ids": [
        12,
        18,
        10751,
        14
      ],
      "backdrop_path": "/eIOTsGg9FCVrBc4r2nXaV61JF4F.jpg",
      "adult": false,
      "overview": "A man-cub named Mowgli fostered by wolves. After a threat from the tiger Shere Khan, Mowgli is forced to flee the jungle, by which he embarks on a journey of self discovery with the help of the panther, Bagheera and the free-spirited bear, Baloo.",
      "release_date": "2016-04-07"
    },
    {
      "vote_count": 3910,
      "id": 47933,
      "video": false,
      "vote_average": 5.1,
      "title": "Independence Day: Resurgence",
      "popularity": 21.224,
      "poster_path": "/5CHJs479xWnm3zMDOl94VkKS7MZ.jpg",
      "original_language": "en",
      "original_title": "Independence Day: Resurgence",
      "genre_ids": [
        28,
        12,
        878
      ],
      "backdrop_path": "/mou66QcacavDahfyuaqKfaCLdoI.jpg",
      "adult": false,
      "overview": "We always knew they were coming back. Using recovered alien technology, the nations of Earth have collaborated on an immense defense program to protect the planet. But nothing can prepare us for the aliens’ advanced and unprecedented force. Only the ingenuity of a few brave men and women can bring our world back from the brink of extinction.",
      "release_date": "2016-06-22"
    },
    {
      "vote_count": 1571,
      "id": 300671,
      "video": false,
      "vote_average": 7,
      "title": "13 Hours: The Secret Soldiers of Benghazi",
      "popularity": 20.266,
      "poster_path": "/4gpW4MkRI0BB17GHOJzXq0WamDP.jpg",
      "original_language": "en",
      "original_title": "13 Hours: The Secret Soldiers of Benghazi",
      "genre_ids": [
        28,
        18,
        36,
        53,
        10752
      ],
      "backdrop_path": "/hY8myWhjSOUeeiPskhOfwwgaCwc.jpg",
      "adult": false,
      "overview": "An American Ambassador is killed during an attack at a U.S. compound in Libya as a security team struggles to make sense out of the chaos.",
      "release_date": "2016-01-13"
    },
    {
      "vote_count": 4708,
      "id": 381284,
      "video": false,
      "vote_average": 8,
      "title": "Hidden Figures",
      "popularity": 19.889,
      "poster_path": "/6cbIDZLfwUTmttXTmNi8Mp3Rnmg.jpg",
      "original_language": "en",
      "original_title": "Hidden Figures",
      "genre_ids": [
        18,
        36
      ],
      "backdrop_path": "/aRP8QAOhnthglsTUzeLWwoPVV1T.jpg",
      "adult": false,
      "overview": "The untold story of Katherine G. Johnson, Dorothy Vaughan and Mary Jackson – brilliant African-American women working at NASA and serving as the brains behind one of the greatest operations in history – the launch of astronaut John Glenn into orbit. The visionary trio crossed all gender and race lines to inspire generations to dream big.",
      "release_date": "2016-12-10"
    },
    {
      "vote_count": 1673,
      "id": 411088,
      "video": false,
      "vote_average": 8.2,
      "title": "The Invisible Guest",
      "popularity": 19.646,
      "poster_path": "/fptnZJbLzKUHeNlYrAynbyoL5YJ.jpg",
      "original_language": "es",
      "original_title": "Contratiempo",
      "genre_ids": [
        9648,
        80,
        53
      ],
      "backdrop_path": "/urrIaUxmLXYWodZx7KJPLLVFaKC.jpg",
      "adult": false,
      "overview": "A young businessman wakes up in a hotel room, locked from the inside, along his lover, who was murdered while he was unconscious. He hires a prestigious lawyer, and over the course of one evening, they must work together to build a defense case for him before he is taken to jail.",
      "release_date": "2016-09-22"
    },
    {
      "vote_count": 3609,
      "id": 324668,
      "video": false,
      "vote_average": 6.1,
      "title": "Jason Bourne",
      "popularity": 19.209,
      "poster_path": "/lFSSLTlFozwpaGlO31OoUeirBgQ.jpg",
      "original_language": "en",
      "original_title": "Jason Bourne",
      "genre_ids": [
        28,
        53
      ],
      "backdrop_path": "/dtFiuavnsnl0krJAeSmGAOFAhw2.jpg",
      "adult": false,
      "overview": "The most dangerous former operative of the CIA is drawn out of hiding to uncover hidden truths about his past.",
      "release_date": "2016-07-27"
    },
    {
      "vote_count": 10381,
      "id": 329865,
      "video": false,
      "vote_average": 7.4,
      "title": "Arrival",
      "popularity": 19.189,
      "poster_path": "/hLudzvGfpi6JlwUnsNhXwKKg4j.jpg",
      "original_language": "en",
      "original_title": "Arrival",
      "genre_ids": [
        18,
        878,
        9648
      ],
      "backdrop_path": "/yIZ1xendyqKvY3FGeeUYUd5X9Mm.jpg",
      "adult": false,
      "overview": "Taking place after alien crafts land around the world, an expert linguist is recruited by the military to determine whether they come in peace or are a threat.",
      "release_date": "2016-11-10"
    }
  ]
}