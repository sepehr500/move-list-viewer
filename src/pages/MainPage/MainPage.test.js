import React from "react";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement
} from "@testing-library/react";
import renderer from "react-test-renderer";
import { MainPage } from "./index";
import fakeResponse from "./fakeResponse.json";

const mock = new MockAdapter(axios);

// When there is more then one request make this onGet more specific
mock.onGet().reply(200, fakeResponse);

const waitForMovieList = getByTestIdFn =>
  waitForElement(() => getByTestIdFn("movie-list"));

// Bit of a hack to wait for the microqueue to empty out
function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

it("renders loading correctly", async () => {
  const { getByText } = render(<MainPage />);
  expect(getByText("Loading...")).toBeTruthy();
});

it("renders data correctly", async () => {
  const tree = renderer.create(<MainPage />);
  await flushPromises();
  expect(tree.toJSON()).toMatchSnapshot();
});

it("renders correct number of cards", async () => {
  const { getByTestId } = render(<MainPage />);
  expect(getByTestId("movie-list").children.length).toBe(20);
});

describe("view details flow", () => {
  beforeEach(() => {
    cleanup();
  });
  it("shows details when details button is clicked", async () => {
    const { getByText, getAllByText, getByTestId } = render(<MainPage />);
    await waitForMovieList(getByTestId);
    fireEvent.click(getAllByText("details")[0]);
    expect(getByText("close")).toBeTruthy();
    expect(getByText("Add to favorites")).toBeTruthy();
  });

  it("closes modal when close is clicked", async () => {
    const { getByText, getAllByText, queryByTestId, getByTestId } = render(
      <MainPage />
    );
    await waitForMovieList(getByTestId);
    fireEvent.click(getAllByText("details")[0]);
    expect(getByText("close")).toBeTruthy();
    fireEvent.click(getByText("close"));
    expect(queryByTestId("close-button")).toBeFalsy();
  });

  it("marks favorite when Add to favorites is clicked", async () => {
    const { getByText, getAllByText, getByTestId } = render(<MainPage />);
    await waitForMovieList(getByTestId);
    fireEvent.click(getAllByText("details")[0]);

    expect(getByText("Add to favorites")).toBeTruthy();
    fireEvent.click(getByText("Add to favorites"));

    expect(getByText("Unfavorite")).toBeTruthy();
    // Not ideal. But no way of knowing that the star is gold without checking the class
    expect(Object.values(getByTestId("star").classList)).toContain("gold");
  });

  it("unfavorites when Unfavorite is clicked", async () => {
    const { getByText, getAllByText, getByTestId } = render(<MainPage />);
    await waitForMovieList(getByTestId);
    fireEvent.click(getAllByText("details")[0]);

    expect(getByText("Add to favorites")).toBeTruthy();
    fireEvent.click(getByText("Add to favorites"));

    expect(getByText("Unfavorite")).toBeTruthy();
    expect(Object.values(getByTestId("star").classList)).toContain("gold");

    fireEvent.click(getByText("Unfavorite"));
    expect(Object.values(getByTestId("star").classList)).toContain("gray");
  });
});
