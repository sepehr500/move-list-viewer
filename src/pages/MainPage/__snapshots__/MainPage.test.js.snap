// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders data correctly 1`] = `
<MainPage>
  <div>
    Loading...
  </div>
</MainPage>
`;
